protocol/x/vault/keeper/withdraw.go

```diff
@@ -117,3 +115,164 @@ func (k Keeper) GetVaultWithdrawalSlippage(
+		return nil, types.ErrOwnerNotFound
+	}
+	ownerShareUnlocks, _ := k.GetOwnerShareUnlocks(ctx, toSubaccount.Owner)
+	ownerSharesAfterWithdrawal := ownerShares.NumShares.BigInt()
+	ownerSharesAfterWithdrawal.Sub(ownerSharesAfterWithdrawal, sharesToWithdraw)
+	ownerLockedShares := ownerShareUnlocks.GetTotalLockedShares()
+	if ownerSharesAfterWithdrawal.Cmp(ownerLockedShares) < 0 {
+		return nil, errorsmod.Wrapf(
+			types.ErrInsufficientWithdrawableShares,
+			"shares to withdraw: %s, owner total shares: %s, owner locked shares: %s",
+			sharesToWithdraw,
+			ownerShares,
+			ownerLockedShares,
+		)
+	}
+
+	// 2. Redeem from main vault.
+	totalShares := k.GetTotalShares(ctx).NumShares.BigInt()
+	redeemedQuoteQuantums, err = k.GetSubaccountEquity(ctx, types.MegavaultMainSubaccount)
+	if err != nil {
+		log.ErrorLogWithError(ctx, "Failed to get megavault main vault equity", err)
+		return nil, err
+	}
+	redeemedQuoteQuantums.Mul(redeemedQuoteQuantums, sharesToWithdraw)
+	redeemedQuoteQuantums.Quo(redeemedQuoteQuantums, totalShares)
+
+	// 3. Redeem from each sub vault.
+	vaultParamsIterator := k.getVaultParamsIterator(ctx)
+	defer vaultParamsIterator.Close()
+	for ; vaultParamsIterator.Valid(); vaultParamsIterator.Next() {
+		vaultId, err := types.GetVaultIdFromStateKey(vaultParamsIterator.Key())
+		if err != nil {
+			log.ErrorLogWithError(ctx, "Failed to get vault ID from state key", err)
+			continue
+		}
+		var vaultParams types.VaultParams
+		k.cdc.MustUnmarshal(vaultParamsIterator.Value(), &vaultParams)
+
+		_, perpetual, marketParam, marketPrice, err := k.GetVaultClobPerpAndMarket(ctx, *vaultId)
+		if err != nil {
+			log.ErrorLogWithError(ctx, "Failed to get perpetual and market", err, "Vault ID", vaultId)
+			continue
+		}
+		leverage, equity, err := k.GetVaultLeverageAndEquity(ctx, *vaultId, &perpetual, &marketPrice)
+		if err != nil {
+			log.ErrorLogWithError(ctx, "Failed to get vault leverage and equity", err, "Vault ID", vaultId)
+			continue
+		}
+
+		slippage, err := k.GetVaultWithdrawalSlippage(
+			ctx,
+			*vaultId,
+			sharesToWithdraw,
+			totalShares,
+			leverage,
+			&perpetual,
+			&marketParam,
+		)
+		if err != nil {
+			log.ErrorLogWithError(
+				ctx,
+				"Failed to get vault withdrawal slippage",
+				err,
+				"Vault ID",
+				vaultId,
+			)
+			continue
+		}
+
+		// Transfer `equity * shares / totalShares * (1 - slippage)` from sub vault to main vault.
+		redeemedFromSubVault := new(big.Rat).SetFrac(equity, big.NewInt(1))
+		redeemedFromSubVault.Mul(redeemedFromSubVault, new(big.Rat).SetFrac(sharesToWithdraw, totalShares))
+		redeemedFromSubVault.Mul(redeemedFromSubVault, new(big.Rat).Sub(lib.BigRat1(), slippage))
+		quantumsToTransfer := new(big.Int).Quo(redeemedFromSubVault.Num(), redeemedFromSubVault.Denom())
+
+		err = k.subaccountsKeeper.TransferFundsFromSubaccountToSubaccount(
+			ctx,
+			*vaultId.ToSubaccountId(),
+			types.MegavaultMainSubaccount,
+			assetstypes.AssetUsdc.Id,
+			quantumsToTransfer,
+		)
```

Comment:
question, why not just transfer from the vault to the destination subaccount rather than all to the main vault first, then the subaccount.
Might be minor, but would reduce gas usage.

Reply:
I guess this is better for the actual transfer events though, instead of a random set of 100+ transfers, the withdrawing user only sees a single transfer.